utils/exception_container.hh

```diff
@@ -0,0 +1,143 @@
+    // Keep an integer which identifies the variant.
+    // Bonus points: if each error type has a unique, globally-assigned
+    // identified integer, then conversion of the exception_container
+    // to a container supporting a superset of errors becomes very cheap.
+    seastar::foreign_ptr<std::unique_ptr<std::variant<Exs...>>> _eptr;
+
+    void check_nonempty() const {
+        if (empty()) {
+            throw bad_exception_container_access();
+        }
+    }
+
+public:
+    // Constructs an exception_container which does not contain any exception.
+    exception_container() = default;
+
+    template<typename Ex>
+    requires VariantElement<Ex, exception_variant>
+    exception_container(Ex&& ex)
+            : _eptr(seastar::make_foreign(std::make_unique<exception_variant>(std::move(ex))))
+    { }
+
+    inline bool empty() const {
+        return __builtin_expect(!_eptr, false);
+    }
+
+    inline operator bool() const {
+        return !empty();
+    }
+
+    // Accepts a visitor
+    auto accept(auto f) const {
+        check_nonempty();
+        return std::visit(std::move(f), *_eptr);
+    }
+
+    // Throws currently held exception as a C++ exception.
+    // If the container is empty, it throws bad_exception_container_access.
+    [[noreturn]] void throw_me() const {
+        check_nonempty();
+        std::visit([] (const auto& ex) { throw ex; }, *_eptr);
+        std::terminate(); // Should be unreachable
+    }
+
+    // Creates an exceptional future from this error.
+    // The exception is copied into the new exceptional future.
+    // If the container is empty, returns an exceptional future
+    // with the bad_exception_container_access exception.
+    template<typename T = void>
+    seastar::future<T> as_exception_future() const & {
+        if (!_eptr) {
+            return seastar::make_exception_future<T>(bad_exception_container_access());
+        }
+        return std::visit([] (const auto& ex) {
+            return seastar::make_exception_future<T>(ex);
+        }, *_eptr);
+    }
+
+    // Transforms this exception future into an exceptional future.
+    // The exception is moved out and the container becomes empty.
+    // If the container was empty, returns an exceptional future
+    // with the bad_exception_container_access exception.
+    template<typename T = void>
+    seastar::future<T> into_exception_future() && {
+        if (!_eptr) {
+            return seastar::make_exception_future<T>(bad_exception_container_access());
+        }
+        auto f = std::visit([] (auto&& ex) {
+            return seastar::make_exception_future<T>(std::move(ex));
+        }, *_eptr);
+        _eptr.reset();
+        return f;
+    }
+};
+
+template<typename... Exs>
+inline std::ostream& operator<<(std::ostream& os, const exception_container<Exs...>& ec) {
+    ec.accept([&os] (const auto& ex) {
+        os << typeid(ex).name() << " (" << ex.what() << ")";
+    });
+    return os;
+}
```

Comment:
Copying my previous comment which got marked "outdated" by github:

I think you should use the existing operator<< from <seastar/utils/log.hh> instead of open-coding your own exception printing code.
One of the problems in your code here is that you print typeid(ex).name() without demangling the ugly C++-mangled names. At the minimum you should be using seastar::pretty_type_name() - but I think you can be using the entire operator<< from Seastar.

Reply:
Answered under the "outdated" comment, and done.